grakn-graph/src/main/java/ai/grakn/graph/internal/InstanceImpl.java

```diff
@@ -61,30 +62,30 @@
  *           For example {@link ai.grakn.concept.EntityType} or {@link RelationType}
  */
 abstract class InstanceImpl<T extends Instance, V extends Type> extends ConceptImpl<T> implements Instance {
-    private ElementCache<TypeLabel> cachedInternalType = new ElementCache<>(() -> {
-        int typeId = getProperty(Schema.ConceptProperty.INSTANCE_TYPE_ID);
-        Type type = getGraknGraph().getConcept(Schema.ConceptProperty.TYPE_ID, typeId);
+    private Cache<TypeLabel> cachedInternalType = new Cache<>(() -> {
+        int typeId = property(Schema.ConceptProperty.INSTANCE_TYPE_ID);
+        Type type = graph().getConcept(Schema.ConceptProperty.TYPE_ID, typeId);
         return type.getLabel();
     });
 
-    private ElementCache<V> cachedType = new ElementCache<>(() -> {
-        ConceptImpl<?> currentType = (ConceptImpl) this.getOutgoingNeighbours(Schema.EdgeLabel.ISA).findFirst().orElse(null);
+    private Cache<V> cachedType = new Cache<>(() -> {
+        Optional<EdgeElement> typeEdge = vertex().getEdgesOfType(Direction.OUT, Schema.EdgeLabel.ISA).
+                flatMap(edge -> edge.target().getEdgesOfType(Direction.OUT, Schema.EdgeLabel.SHARD)).findAny();
 
-        while (currentType.isShard()){
-            currentType = (ConceptImpl) currentType.getOutgoingNeighbours(Schema.EdgeLabel.SHARD).findFirst().orElse(null);
+        if(!typeEdge.isPresent()) {
+            throw GraphOperationException.noType(this);
```

Comment:
I think this might be another "unreachable statement"-style error. We should probably talk about how to handle these in general.